### ISSUE TITLE
Initial Setup

### ISSUE THREAD

**Issue by alphafox02** [Xinyuan-LilyGO/T-Halow#1]
Good day. I received the T-Halow units today, but I have a few questions after looking over your site. 

When I plug in via USBC on a Linux PC I can see the ttyACM0 and go there to see stats about the state of the various interfaces followed by disconnect, this repeats as long as I'm connected.

It then seems like I need to load a TX firmware to the Halow radio, it mentions doing so by running various AT commands, but doesn't fully explain the process. If the T-Halow is plugged into internet is it able to go out and grab firmware? Does the firmware need to be downloaded on the PC and then somehow be put over onto the T-Halow? Do I need the programmer it talks about in the README?

I tried attaching a microUSB cable when opens the ttyUSB0 interface, but it to seems read only with details spitting out every so often.

I'm just looking to set one up as a AP and one as a STA. 

Thank you.

**Comment by irongiant33** [comment]
confirm I'm experiencing the same issue. 

Below is the serial output from what I only assume is the ESP32-S3, connected via USB-C. This process repeats indefinitely as long as the device is connected.

```
[  3114][I][esp32-hal-i2c.c:75] i2cInit(): Initialising I2C Master: sda=7 scl=6 freq=100000
OLED FeatherWing test
E (3040) camera: Camera probe failed with error 0x105(ESP_ERR_NOT_FOUND)
Camera init failed with error 0x105[  3149][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3250][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3350][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3450][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x00
[  3450][W][sd_diskio.cpp:516] ff_sd_initialize(): GO_IDLE_STATE failed
[  3451][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  3460][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3565][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3666][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3767][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x00
Card Mount Failed
OK
AT+SYSDBG SUCCEED

OK
AT+BSS_BW SUCCEED

OK
AT+MODE=STA SUCCEED
******************************
TX-AH   PASS
SSD1306  ---
SDCard   ---
CAMERA   ---
 
Role:          STA 
Disconnect!!!

+DISCONNECT
******************************
TX-AH   PASS
SSD1306  ---
SDCard   ---
CAMERA   ---
 
Role:          STA 
Disconnect!!!

[271196]send assoc request to 1e:05:59:63:77:78, ssid= (freq:9240)
[271457]send assoc request to 1e:05:59:63:77:78, ssid= (freq:9080)
[271673]send assoc request to 1e:05:59:63:77:78, ssid= (freq:9080)
[273550]send assoc request to 1e:05:59:63:77:78, ssid= (freq:9160)
+DISCONNECT
******************************
TX-AH   PASS
SSD1306  ---
SDCard   ---
CAMERA   ---
 
Role:          STA 
Disconnect!!!

[274555]send assoc request to 1e:05:59:63:77:78, ssid= (freq:9160)
[276189]send assoc request to 1e:05:59:63:77:78, ssid= (freq:9240)
+DISCONNECT
******************************
```

I am unable to send any serial commands over this interface. 

When I connect to the T-HaLow chip via micro-USB, I see a USB interface pop up. I can connect to it, but there is no log output like there is for the ESP32 and I also cannot send any information to the device over serial.

Lastly, I tried connecting an ethernet cable to see if I get an IP address, but I am not assigned one via DHCP. Do I need to statically assign an IP address in order to communicate with the T-HaLow?

**Comment by alphafox02** [comment]
On the micro USB I get output showing statistics and info, I’ll have to look again though. On the USBC connection I don’t recall ever seeing a pass next to the TX-AH, but I’ve also not switched out the chip in the T HaLow holder with the one in the bag. It’s not clear to me if it’s flashed yet or not and/or how to flash the AP vs STA firmware.

**Comment by irongiant33** [comment]
Furthermore, does the 32MBit serial flash SS package [Winbond 25Q32JVSIQ](https://www.winbond.com/hq/product/code-storage-flash-memory/serial-nor-flash/?__locale=en&partNo=W25Q32JV) provide memory for the T-HaLow, the ESP32 or both? I thought ESP32s came with their own native flash, in the case of the S3, 8MB of it. I used the following command to read & store a flash backup in case I mess things up later:

```
python3 esptool.py --chip esp32s3 --port /dev/ttyACM0 --baud 115200 read_flash 0 0x800000 thalow-flash-contents.bin
```

Then I tried to flash the AP firmware to the ESP32 because it ships with the STA by default. Since I have 2x T-HaLows, I thought maybe since having one as the AP and one as the STA, it might encourage a connection. I used [esptool](https://github.com/espressif/esptool) for this and downloaded the firmware from this repository.

```
python3 esptool.py --chip esp32s3 --port /dev/ttyACM0 --baud 115200 write_flash 0 ~/Downloads/firmware_T-Halow_AP_240628.bin
```

I confirmed that the mode was changed to AP, but there was still no apparent packet connections or exchange between my two devices.

```
[  3114][I][esp32-hal-i2c.c:75] i2cInit(): Initialising I2C Master: sda=7 scl=6 freq=100000
OLED FeatherWing test
E (3040) camera: Camera probe failed with error 0x105(ESP_ERR_NOT_FOUND)
Camera init failed with error 0x105[  3149][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3250][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3350][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3449][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x00
[  3450][W][sd_diskio.cpp:516] ff_sd_initialize(): GO_IDLE_STATE failed
[  3451][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  3460][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3565][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3666][W][sd_diskio.cpp:174] sdCommand(): no token received
[  3767][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x00
Card Mount Failed
OK
AT+SYSDBG SUCCEED

OK
AT+BSS_BW SUCCEED

OK
AT+MODE=AP SUCCEED
******************************
TX-AH   PASS
SSD1306  ---
SDCard   ---
CAMERA   ---
 
Role:           AP 
Disconnect!!!

+DISCONNECT
******************************
TX-AH   PASS
SSD1306  ---
SDCard   ---
CAMERA   ---
 
Role:           AP 
Disconnect!!!

+DISCONNECT
******************************
TX-AH   PASS
SSD1306  ---
SDCard   ---
CAMERA   ---
 
Role:           AP 
Disconnect!!!

+DISCONNECT
******************************
```

**Comment by irongiant33** [comment]
Lastly, after connecting the TF card there is no longer a verbose output like above. I connected a SSD1306 to 3V3, GND, SCL to pin 6 and SDA to pin 7. I did not get a "PASS" next to the SSD1306 on the debug log, nor did I see any information displayed on the screen. I thought this was the correct place to put the SSD1306?

**Comment by alphafox02** [comment]
After using the one application in the README to send commands and sync, I have one node in AP mode and one in STA with a connection between the two. 

I have no idea what the chip in each by default did. But it seems like the flash chip in the bag must have been programmed at the factory with the bin file because it’s not until after inserting it that the halow radio becomes usable.

I now do not understand how to enable the Ethernet port. In one section of the readme sit shows a diagram of mode 1 and mode 2 and says to acheive mode 2 to not download any firmware on the esp. I do not understand what that means. In another image online, it mentions the Ethernet can be enabled by flashing firmware and it points to the replaceable chip. 

Is the Ethernet not enabled from the factory? I’m mostly interested in being able to plug a computer into a AP and one into a STA T-HaLow, assign a static IP, and ping one another. I’ve not been able to achieve that yet.

**Comment by irongiant33** [comment]
Agree with all observations from @alphafox02 above. I get the following output after issuing a `AT+` command via the Arduino Serial Monitor with the "Adafruit Feather ESP32-S3 No PSRAM" as my board.

```
0. AT+ACS_START= 
 1. AT+ACK_TO= 
 2. AT+ADC_DUMP= 
 3. AT+AP_SLEEP_MODE= 
 4. AT+ANT_DUAL= 
 5. AT+ANT_CTRL= 
 6. AT+ANT_AUTO= 
 7. AT+ANT_DEF= 
 8. AT+BGRSSI_MARGIN= 
 9. AT+BGRSSI_MAX= 
10. AT+BGRSSI_SPUR= 
11. AT+BSS_BW= 
12. AT+BUS_WT= 
13. AT+CCA_OBSV= 
14. AT+CCA_CE= 
15. AT+CCMP_SUPPORT= 
16. AT+CHAN_SCAN= 
17. AT+CS_CNT= 
18. AT+CS_EN= 
19. AT+CS_NUM= 
20. AT+CS_PERIOD= 
21. AT+CS_TH= 
22. AT+CTS_DUP= 
23. AT+EDCA_AIFS=
24. AT+EDCA_CW= 
25. AT+EDCA_TXOP= 
26. AT+EVM_MARGIN= 
27. AT+FREQ_LIST= 
28. AT+FT_ATT= 
29. AT+LMAC_DBGSEL= 
30. AT+LO_FREQ= 
31. AT+LOADDEF= 
32. AT+MAC_ADDR= 
33. AT+MCAST_DUP= 
34. AT+MCAST_REORDER= 
35. AT+MCAST_BW= 
36. AT+MCAST_MCS= 
37. AT+MCAST_RTS= 
38. AT+NOR_RD= 
39. AT+OBSS_CCA_DIFF= 
40. AT+OBSS_EDCA= 
41. AT+OBSS_NAV_DIFF= 
42. AT+OBSS_SWITCH= 
43. AT+OBSS_TH= 
44. AT+OBSS_PER= 
45. AT+PCF_EN= 
46. AT+PCF_PERCENT= 
47. AT+PCF_PERIOD= 
48. AT+PHY_RESET= 
49. AT+PRI_CHAN= 
50. AT+PRINT_PERIOD= 
51. AT+QA_ATT= 
52. AT+QA_CFG= 
53. AT+QA_RESULTS= 
54. AT+QA_RXTHD= 
55. AT+QA_START= 
56. AT+QA_TXTHD= 
57. AT+RC_NEW= 
58. AT+REG_RD= 
59. AT+REG_WT= 
60. AT+RF_RESET= 
61. AT+RTS_DUP= 
62. AT+RX_AGC= 
63. AT+RX_ERR= 
64. AT+RX_EVM= 
65. AT+RX_PKTS= 
66. AT+RX_REORDER= 
67. AT+RX_RSSI= 
68. AT+SET_AGC= 
69. AT+SET_AGC_TH= 
70. AT+SET_BGRSSI= 
71. AT+SET_BGRSSI_AVG= 
72. AT+SET_RTS= 
73. AT+SET_VDD13= 
74. AT+SHORT_GI= 
75. AT+SHORT_TH= 
76. AT+SLEEP_EN= 
77. AT+T_SENSOR= 
78. AT+TEST_START= 
79. AT+TX_AGG_AUTO= 
80. AT+TX_ATTN= 
81. AT+TX_BW= 
82. AT+TX_BW_DYNAMIC= 
83. AT+TX_CNT_MAX= 
84. AT+TX_CONT= 
85. AT+TX_DELAY= 
86. AT+TX_DST_ADDR= 
87. AT+TX_FAIL= 
88. AT+TX_FC= 
89. AT+TX_FLAGS= 
90. AT+TX_LEN= 
91. AT+TX_MAX_AGG= 
92. AT+TX_MAX_SYMS= 
93. AT+TX_MCS= 
94. AT+TX_MCS_MAX= 
95. AT+TX_MCS_MIN= 
96. AT+TX_ORDERED= 
97. AT+TX_PHA_AMP= 
98. AT+TX_PKTS= 
99. AT+TX_PWR_AUTO= 
100. AT+TX_PWR_MAX= 
101. AT+TX_PWR_SUPER= 
102. AT+TX_PWR_SUPER_TH= 
103. AT+TX_RATE_FIXED= 
104. AT+TX_START= 
105. AT+TX_STEP= 
106. AT+TX_TRIG= 
107. AT+TX_TYPE= 
108. AT+TXOP_EN= 
109. AT+WAKE_EN= 
110. AT+XO_CS= 
111. AT+XO_CS_AUTO= 
112. AT+LO_TABLE= 
113. AT+PS_CHECK= 
114. AT+RADIO_ONOFF= 
115. AT+STA_INFO= 
```

I would expect the `AT+PAIR=1` and `AT+PAIR=0` to be a part of this, but it isn't. I was able to get my devices to pair by issuing both the `AT+PAIR=1` command. Is there somewhere where all of the AT commands for the T-HaLow are documented to enable what @alphafox02 mentioned in terms of getting an IP address assigned to each connected to a computer for pinging across a HaLow network?

**Comment by Duan-GX** [comment]
@alphafox02 
You can find more information about the TX-AH module here：
  https://en.taixin-semi.com/Product?prouctSubClass=33
![Snipaste_2024-07-30_09-53-18](https://github.com/user-attachments/assets/8bc1e01f-41b1-4dd7-a714-0dfdbbb436c3)


To use your computer, insert T-Halow and Ping it. Need to do:
1、Download the 'huge-ic-ah_v1.6.3.3-26347_2023.11.29_TAIXIN-WNB.bin' firmware to AH-TX;
![image](https://github.com/user-attachments/assets/0c1723b9-5129-4268-9a3a-e7b3c1635505)
2、and then the wiring according to the following way;
(1). Used as power supply
(2). Used to send and receive AT instructions
(3). Connect the computer to the network cable
![image](https://github.com/user-attachments/assets/0631a770-4fe8-4002-82e6-cbecce22f5ea)
3、Open the serial port AT position 2, and then send the 'AT+' command in the serial port assistant, as shown in the following figure. If there is a response, it means that the firmware download of 'huge-ic-ah_v1.6.3.3-26347_2023.11.29_TAIXIN-WNB.bin' is successful.
![image](https://github.com/user-attachments/assets/d46e6656-5957-46ca-9012-6221577674a9)
4、Turn off the computer's firewall, open the computer's "Network and sharing center", find a network cable to connect T-Halow to the computer, and you will see an unidentified network in the "Network and Sharing Center"
5、Set an IP address for T-Halow
![image](https://github.com/user-attachments/assets/1f87ed03-2928-4b4d-9b56-baf2ced6a902)
6、Finally, Ping the address you just set from the command line window
![image](https://github.com/user-attachments/assets/e1c19823-33f2-41d4-9a59-1d2fa9611f0d)

**Comment by Duan-GX** [comment]
@irongiant33 
The following address can consult TX-AH more information, may be helpful to you；
https://en.taixin-semi.com/Product?prouctSubClass=33
You can find the AT instructions recorded in the <泰芯AH模组AT指令开发指南>

**Comment by irongiant33** [comment]
> ```
> python3 esptool.py --chip esp32s3 --port /dev/ttyACM0 --baud 115200 write_flash 0 ~/Downloads/firmware_T-Halow_AP_240628.bin
> ```

confirmed response from `AT+` command on both the AP and STA. Set the STA for an IP of 10.10.10.156 and the AP with 10.10.10.157. I could ping from each respective host to the T-HaLow, but I could not send data across the network.

```
$ ping 10.10.10.156
PING 10.10.10.156 (10.10.10.156) 56(84) bytes of data.
From 10.10.10.157 icmp_seq=1 Destination Host Unreachable
From 10.10.10.157 icmp_seq=2 Destination Host Unreachable
From 10.10.10.157 icmp_seq=3 Destination Host Unreachable
From 10.10.10.157 icmp_seq=4 Destination Host Unreachable
From 10.10.10.157 icmp_seq=5 Destination Host Unreachable
From 10.10.10.157 icmp_seq=6 Destination Host Unreachable
From 10.10.10.157 icmp_seq=7 Destination Host Unreachable
From 10.10.10.157 icmp_seq=8 Destination Host Unreachable
From 10.10.10.157 icmp_seq=9 Destination Host Unreachable
^C
--- 10.10.10.156 ping statistics ---
11 packets transmitted, 0 received, +9 errors, 100% packet loss, time 10269ms
pipe 4

$ ping 10.10.10.156
PING 10.10.10.156 (10.10.10.156) 56(84) bytes of data.
From 10.10.10.157 icmp_seq=1 Destination Host Unreachable
From 10.10.10.157 icmp_seq=2 Destination Host Unreachable
```

**Comment by alphafox02** [comment]
When you say download this file,
1、Download the 'huge-ic-ah_v1.6.3.3-26347_2023.11.29_TAIXIN-WNB.bin' firmware to AH-TX;"

I don't have the flashing tool that is shown in the pictures. I took the chip out of the bag that says flash and installed it into the T-Halow. With that chip I can run the AT commands. 

You're saying to set an IP Address for the T-Halow, but what I'm wanting to do is set an IP address for the computer. I don't want to just ping the T-Halow, I want to ping across the T-Halow network to the computer on the other T-Halow's ethernet port. 

Maybe I'm not understanding the capabilities of the device. I'm also on Linux. If I plug from the T-Halow to my machine, even in Windows, when setting the Ip as shown in your diagram and pinging it, you're just pinging yourself it seems?

**Comment by alphafox02** [comment]
Can you confirm what firmware is loaded onto the chip that comes in the bag? Is it this firmware? If it's not that firmware, do I need the flashing programming tool to load it?

huge-ic-ah_v1.6.3.3-26347_2023.11.29_TAIXIN-WNB.bin

What does the WNB stand for? Does it mean wireless network bridge?

**Comment by alphafox02** [comment]
Okay. I'm understanding that the flash chip in the bag does in fact contain firmware with Ethernet functions and that WNB does mean bridge. So now I just need to see how it's possible to have network traffic flow across the bridge from one PC to another.

**Comment by irongiant33** [comment]
> Okay. I'm understanding that the flash chip in the bag does in fact contain firmware with Ethernet functions and that WNB does mean bridge. So now I just need to see how it's possible to have network traffic flow across the bridge from one PC to another. 

As @alphafox02 mentions, is there not a static IP address or DHCP functionality in the WNB firmware? Because assigning a static IP in the windows/Linux network manager is just assigning an IP on your local computer and not the T-HaLow. It would make sense if there was DHCP or a static IP address that could be reconfigured in order to get multiple devices to communicate in a mesh format

**Comment by irongiant33** [comment]
Also the flash programmer you mentioned is very expensive for hobbyists. After some basic research it seems like something like this might work as an inexpensive option? In case we want to flash the Winbond serial flash again

KeeYees SOP8 SOIC8 Test Clip and CH341A USB Programmer Flash for Most of 24 25 Series BIOS Chip with PDF Tutorial https://a.co/d/0QmnWVG

@Duan-GX

**Comment by alphafox02** [comment]
Normally in a wireless bridge, you actually do want to just assign an IP on your computer. I assumed the T-Halow is running in such a way that any network traffic can pass over it, meaning I'd need to have an IP address assigned to whatever device is plugged into it's ethernet port. 

Looking at the readme again, it appears there's two methods listed. 1.1 and 1.2.

In 1.1 it talks about the WNB firmware. It's not until 1.2 does it talk about loading another firmware and setting up one as an AP and one as a STA.

Are they two distinct modes? How in 1.1 when using the WNB firmware are two T-Halows communicating with one another? Do they just link up and act simply as a Bridge for one device plugged into one to talk to another device plugged into the other, so long as static IPs are set on the devices themselves (not actually the T-Halow, but the device plugged in). 

If 1.1 is the mode that's to be used with WNB firmware, how do you get the ESPs out of AP and STA mode if they've already been loaded with firmware?

**Comment by ShallowGreen123** [comment]
You can see if the following connection is helpful for you：
https://github.com/Xinyuan-LilyGO/T-Halow/tree/master/hardware/TX_AH%20tool

The TX-AH module provides very little information, so we need to explore constantly

**Comment by alphafox02** [comment]
That document and the readme leads me to believe that two T-HaLow modules, on configured to AP and one configured to STA both paired together with AT commands, assuming WNB firmware is loaded on the flash for the Ethernet port to be active, should allow a Pc plugged into each one the ability to reach each other.

Unfortunately that’s not what I’m seeing. I have each T-HaLow configured with the flash chip that was included in the bag. Each T-HaLow correctly responds to AT commands and I have them paired together using the process done using AT commands. I see them each saying connected with RSSI values with the AP mentioning “send_data= with numbers” and the STA saying “RX_data”.

I have a PC plugged into each T-HaLow. One with a static address of 10.10.10.156 and the other with 10.10.10.123. They can’t ping or talk to one another.

**Comment by kylemallory** [comment]
@alphafox02 I just received my own pair of the lilygo modules,  but have yet to configure them.  I intend to perform a similar bridge as you.  You are correct that you should not need an IP address on each HaLow device for bridging.

A proper bridge will forward all data received on the ethernet port to the HaLow and vice-versa.  In theory, you should be able to have 3+ devices with one in AP mode and the other in STA mode, and have all of them talk to each other as though they were all connected via a slow ethernet hub... in theory.

Do you have any indication of data being received on the PHY interface and being sent to the HaLow interface?  A failure of communication could be at any of 3 points:
- the PHY not functioning to decode ethernet frames (possible Wiznet firmware/config)
- the Taixin firmware not receiving data from the PHY (flash chip?)
- the 802.11ah pairing between HaLow devices

Any of these on either end of the HaLow link (or their return path) could be causing the lack of communication, which really means you have 6 points of failure (assuming good hardware).   Ideally,  there should be a way to confirm each stage of the link, but it sounds like the documentation here maybe falling short.

I'm hoping to get mine setup here in the next day or two,  and hopefully I can provide more useful information to help get us all up and running.

**Comment by alphafox02** [comment]
I’ll give a better response here soon. I’m just making my way back from DEFCON. My understanding is unfortunately I need to replace the flash on my chips. I promise I tried every way possible while there are to include soldering chips on temporary boards and removing them. I’m now waiting on some chips to arrive from the company. My hope is pop them in and then the Wireless bridge will work like you and I are on the same page about.

**Comment by jmarcelino** [comment]
I have it working as a bridge but needed to flash the updated WNB firmware with a flash programmer. 

Not tried static IP but configured so the computer connected to the STA end requests DHCP for
 the computer connected to the side. 

Definitely recommend getting a flash programmer of some sort if you want to play with these as you ll need to write the new firmware to the flash chip for any update.

**Comment by kylemallory** [comment]
@alphafox02 @jmarcelino This matches my conclusions as well.  The "stock" flash chip that was pre-installed doesn't appear to have any bin installed (hence TX-AH not reporting "PASS").  Installing the provided flash in the zip-lock allows connecting/communicating via the ESP32.  Per the start-up instructions, it looks like this is the `*TAIXIN-usb.bin`, which doesn't support the "mode 2" setup for PHY connection.  I found a cheap ($14) SOP8 programmer on Amazon that will be here Tuesday, and I'll try it out to flash the pre-installed/blank flash chip with the WNB firmware - it had positive reviews, so hopefully it will prove to be winner.

**Comment by alphafox02** [comment]
I unfortunately didn’t have a programmer on hand. I was having help desolder chips from other Lilygo equipment soldering on the flash chips flashing it And removing again. Can you confirm though that the new firmware they updated just within the last few days that seems like it’s actually dated older work? That’s what you flashed? I thought for sure my method was gonna work, but when I put the chip back in the seat, I would never get the pass so I just packed them up.

**Comment by irongiant33** [comment]
> @alphafox02 @jmarcelino This matches my conclusions as well.  The "stock" flash chip that was pre-installed doesn't appear to have any bin installed (hence TX-AH not reporting "PASS").  Installing the provided flash in the zip-lock allows connecting/communicating via the ESP32.  Per the start-up instructions, it looks like this is the `*TAIXIN-usb.bin`, which doesn't support the "mode 2" setup for PHY connection.  I found a cheap ($14) SOP8 programmer on Amazon that will be here Tuesday, and I'll try it out to flash the pre-installed/blank flash chip with the WNB firmware - it had positive reviews, so hopefully it will prove to be winner.

Any luck with the programmer?

**Comment by kylemallory** [comment]
No, understanding.  It totally burned up the chip (and possibly the programmer)!  I am having to order a new one, and a new flash chip.  I'm still on the hunt for a <$50 programmer that actually supports that flash chip.

**Comment by MetricMoose** [comment]
Back in January I had suggested LilyGo make something with the TX-AH chips and was pretty excited to order a pair in, but I'm not having much luck with getting them setup the way I expected them to work...

The factory software on the pre-installed chip saw the TX-AH (Had PASS next to it). I already have a couple Halow bridges with the TX-AH chipsets, and managed to get the T-Halow paired with my AP using the AT+PAIR commands. So I assume it's setup for "Mode 1" since I'm able to send AT commands to it from the ESP.

I was expecting that "Mode 1" would allow the ESP to get network connectivity through the TX-AH over ethernet based on the diagram of the two modes, though this doesn't seem to be the case unless I'm missing some configuration on the ETH.h library, since I'm not seeing it pull an IP on my network and the ethernet status is down. 

Judging by the "Mode 1 Test" docs, it seems like Mode 1 is only intended for the serial data commands? To me, that _completely_ defeats the purpose of having a network capable MCU attached to a 802.11ah WiFi chipset capable of useful throughput... With the camera attachment, this could have been a really useful devboard for a Halow camera, or other types of portable gadgets that could make use of the long range of Halow, like walkie talkies, weather stations, e-ink displays, ect. Instead, to continue on with my intended project, I'll probably end up attaching a WT32-ETH01 to the T-Halow in WNB mode...

I did manage to get "Mode 2" / WNB working, where it's just acting as a network bridge. I popped in the flash chip from the bag and connected to the MicroUSB serial port to see all the log messages. To connect to my existing HaLow network bridge, I once again ran the the pairing process on both sides (ATH+PAIR=1) until I was getting successful pairing messages on the client, and then put in AT+PAIR=0 on both sides to turn off pairing. I had to also run AT+TX_BW=8MHz. It's still a bit mind boggling that the ESP32 has no control over the TX-AH in this mode, where it could be useful to run a local WiFi/Bluetooth interface for changing settings and getting the status of the TX-AH.

**Comment by alphafox02** [comment]
I’m still waiting on my replacement chips to enable WB mode, but good to hear it’s working for you. I suspect or at least hope maybe future hardware and software revisions will address what you’ve said. I’ve been making suggestions to them as well. 

I’m mainly wanting to plug in a PC on one and reach box plugged in remote on another. Being able to push button pair and/or have a web interface to configure some basic settings would be extremely useful.

**Comment by alphafox02** [comment]
New chips arrived. Network bridge mode is now alive! I’ll share more details when it’s not so late.

**Comment by Humancell** [comment]
Hello!  Thank you for all of the experimentation that you have done so far!  I'm trying to understand the various documentation and how the default board is configured.

1. When shipped, is the default configuration Mode 1 or Mode 2?  I am guessing Mode 1 since the default ESP32 sketch (TX_AH_Rx00p_STA.ino) seems to be running.
2. The USB-C port () is then connected to the ESP32.  Has anyone reprogrammed the ESP32 successfully at this point?
3. The Micro-USB appears to be connected to the HaLow module, and there is output from the various commands being sent to it by the ESP32.  BUT ... this does not appear to be correct since I am not seeing the AT Commands in this output.  So it could be that echo is turned off?

I have a HaLow access point that is up and running - one I built from a Newracom dev kit - and it's set with a SSID and Password.  What I want to do is simply get an AT Command passthrough to the T-HaLow going so that I can attempt to directly control the HaLow module on the T-HaLow to scan for and hopefuly connect to my AP.

At that point I'd be trying to create a ESP32 library to see if I can perform an HTTP POST from the T-HaLow through the Newracom AP to my cloud server.

Thoughts/Feedback/Suggestions?

**Comment by Humancell** [comment]
Some updates to my testing.  When connected to the Micro-USB I am able to send the various AT Commands to the Tianxin HaLow module and they are being handled, even while the ESP32 code is running.

I was able to change the SSID and Password and can see the changes reflected in the output.

08:18:07.365 -> [1893508][31msend assoc request to ff:ff:ff:ff:ff:ff, ssid=Wovyn-HaLow (freq:9080)
08:18:07.365 -> [0m[1894400][31msend assoc request to ff:ff:ff:ff:ff:ff, ssid=Wovyn-HaLow (freq:9160)
08:18:08.299 -> [0m[1894849][31msend assoc request to ff:ff:ff:ff:ff:ff, ssid=Wovyn-HaLow (freq:9160)
08:18:08.707 -> [0m[1895754][31msend assoc request to ff:ff:ff:ff:ff:ff, ssid=Wovyn-HaLow (freq:9240)
08:18:09.615 -> [0m[1897484][31msend assoc request to ff:ff:ff:ff:ff:ff, ssid=Wovyn-HaLow (freq:9240)
08:18:11.374 -> [0m[1897540][31msend assoc request to ff:ff:ff:ff:ff:ff, ssid=Wovyn-HaLow (freq:9080)

I have not been able to get a successful association yet and am exploring the other parameters.

Also, for anyone else wanting the AT Commands in English I did a quick Google translation.  :-)
[English Translation - Taixin AH Module AT instruction development guide_20230524100503.pdf](https://github.com/user-attachments/files/16995204/English.Translation.-.Taixin.AH.Module.AT.instruction.development.guide_20230524100503.pdf)

**Comment by WifiHero91** [comment]
> New chips arrived. Network bridge mode is now alive! I’ll share more details when it’s not so late.

Hello! I'm in the same shoe as You! How i can ask for new chips? 

Thank you!

**Comment by alphafox02** [comment]
I have a few extra I can send you, otherwise Lilygo will probably help you out.

**Comment by WifiHero91** [comment]
@alphafox02 You are very kind! Lilygo has to fix this (i hope) :)

**Comment by Humancell** [comment]
> I have a few extra I can send you, otherwise Lilygo will probably help you out.

If I have a programmer, is the code available somewhere for programming myself?

**Comment by alphafox02** [comment]
The guides LilyGO provide point you to the latest Wireless Bridge firmware and steps to flash, although I think the more recent directions show methods of temporarily soldiering the chips on to another device for flashing. If you look back at the commits and go to earlier ones it shows how to use a programmer. I’ve not seen anyone pop in here saying they managed to get chips flashed by a programmer though.

**Comment by abu-matterize** [comment]
Hi. I'm facing the similar issue, regarding the setup. I've powered it, seeing solid green light. But failing in responding to the `AT` commands when using microUSB port to communicate.

**Comment by abu-matterize** [comment]
I've created a separate issue #14 for this as well.
